egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh

```diff
@@ -0,0 +1,155 @@
+    lattice-prune --acoustic-scale=$acwt --beam=$beam \
+    "ark:gunzip -c $indir/lat.JOB.gz|" ark:- \| \
+    lattice-expand --acoustic-scale=$acwt --epsilon=$epsilon \
+    ark:- "ark:|gzip -c>$expansion_dir/lat.JOB.gz" || exit 1;
+fi
+
+if [ $stage -le 1 ]; then
+  # Convert a lattice into a minimal list of hypotheses under 2 constraints:
+  # 1) Every arc on the lattice must be covered by the list.
+  # 2) Each hypothesis is the best one for at least one arc it includes.
+  echo "$0: converting each lattice into a minimal list of hypotheses."
+  $cmd JOB=1:$nj $dir/log/path_cover.JOB.log \
+    lattice-path-cover --acoustic-scale=$acwt --word-symbol-table=$oldlang/words.txt \
+    "ark:gunzip -c $expansion_dir/lat.JOB.gz|" \
+    ark,t:$paths/path.JOB.words ark,t:$paths/states.JOB ark,t:$paths/path_costs.JOB || exit 1;
+  for n in $(seq $nj); do
+    utils/int2sym.pl -f 2- $oldlang/words.txt < $paths/path.$n.words > $paths/path.$n.words_text || exit 1;
+  done
+fi
+
+if [ $stage -le 2 ]; then
+  echo "$0: computing neural LM scores of the minimal list of hypotheses."
+  $cmd JOB=1:$nj $dir/log/compute_sentence_scores.JOB.log \
+    PYTHONPATH=steps/pytorchnn python steps/pytorchnn/compute_sentence_scores.py \
+     --infile $paths/path.JOB.words_text \
+     --outfile $scores/${model_type}_scores.JOB \
+     --vocabulary $vocab \
+     --model-path $nn_model \
+     --model $model_type \
+     --emsize $embedding_dim \
+     --nhid $hidden_dim \
+     --nlayers $nlayers \
+     --nhead $nhead \
+     --oov "$oov_symbol"
+fi
+
+if [ $stage -le 3 ]; then
+  echo "$0: estimating neural language model scores for each arc."
+  $cmd JOB=1:$nj $dir/log/estimate_arc_nnlm_scores.JOB.log \
+    python steps/pytorchnn/estimate_arc_nnlm_scores.py \
+      --arc-ids $paths/states.JOB \
+      --scores $scores/${model_type}_scores.JOB \
+      --outfile $scores/arc_scores.JOB
+fi
+
+# Rescore the expanded lattice: add neural LM scores first and then remove the
+# old N-gram LM scores. The two models are effectively interpolated.
+oldlm_command="fstproject --project_output=true $oldlm |"
+oldlm_weight=$(perl -e "print -1.0 * $weight;")
+nnlm_weight=$(perl -e "print $weight;")
+if [ $stage -le 4 ]; then
+  echo "$0: replaceing old N-gram LM scores with estimated neural language model scores."
+  if [ "$oldlm" == "$oldlang/G.fst" ]; then
+    $cmd JOB=1:$nj $dir/log/nnlmrescore.JOB.log \
+      lattice-add-nnlmscore --lm-scale=$nnlm_weight "ark:gunzip -c $expansion_dir/lat.JOB.gz|" \
+      $scores/arc_scores.JOB ark:- \| \
+      lattice-lmrescore --lm-scale=$oldlm_weight \
+      ark:- "$oldlm_command" "ark:|gzip -c>$dir/lat.JOB.gz" || exit 1;
+  else
+    $cmd JOB=1:$nj $dir/log/nnlmrescore.JOB.log \
+      lattice-add-nnlmscore --lm-scale=$nnlm_weight "ark:gunzip -c $expansion_dir/lat.JOB.gz|" \
+      $scores/arc_scores.JOB ark:- \| \
+      lattice-lmrescore-const-arpa --lm-scale=$oldlm_weight \
```

Comment:
What is the benefit of doing this versus just taking the best of the n-best-list elements?

Reply:
I did this instead of only taking the neural LM scores since usually interpolating the neural LM with the original n-gram model gives better WERs.

Reply:
.. I mean what is the observed difference in WER?

Reply:
This gives an absolute 0.2% WER improvement on SWBD compared to w/o it.

Reply:
...and no difference on WSJ's eval92 test set. WER on eval92 after lattice rescoring is already very low (1.58% with a Transformer LM). That's probably why interpolating with the n-gram LM doesn't further help.